### ISSUE TITLE
redis.clients.jedis.BinaryJedisCluster : Request to add feature to get reference of ConnectionHandler from JedisClusterConnection.

### ISSUE THREAD

**Issue by saurabhygk** [xetorthio/jedis#1527]
Before starting thing, I want to clear that this not an issue but DETAILED request for new feature
### Expected behavior

I am expecting reference of Connection Handler to get native connection Jedis by implementing some method in either redis.clients.jedis.BinaryJedisCluster or redis.clients.jedis.JedisCluster classes.

### Actual behavior

In actual behaviour, We want to implement pipeline in redis 3.2 cluster but spring-data-redis JedisClusterConnection does not supports the pipeline on cluster. So we planned to create our own by getting native jedis connection from JedisPool. So for that, in current version JedisCluster class do not have any feature which will return us the reference of connection handler. so for that we require to overload BinaryJedisCluster class in our application with same package name and provide some method which returns connectionHandler reference object. But this is not good choice in the scenario if Jedis client version changed in future and some changes implemented in BinaryJedisCluster class then we need to again overload that class and provide that method. Also this approach/solution gives the issue of class loading. Sometime ClassLoader loads the Jedis BinaryJedisCluster class instead of our application overloaded class.

### Steps to reproduce:
This is not issue but DETAILED request. To implement pipeline in our application we need connection handler to get native connection from JedisPool. So for that we need to following steps.
1.  We need to overload BinaryJedisCluster class with package name redis.clients.jedis in our application. We need to implemented following method
`
	public JedisSlotBasedConnectionHandler getConnectionHandler() {
		return (JedisSlotBasedConnectionHandler) this.connectionHandler;
        }
`

2. After that in code where we want Jedis native connection, we have call by following way

// Get Jedis Cluster Connection
``
JedisClusterConnection jedisClusterConnection = (JedisClusterConnection) 
stringRedisTemplate.getConnectionFactory().getClusterConnection();
``
// Get Jedis Cluster reference from Native Connection.
``
JedisCluster jedisCluster = jedisClusterConnection.getNativeConnection();
``
// This method call is new method which we have implemented to get Connection Handler reference from overloaded method, because Jedis library not provides it.
``
JedisSlotBasedConnectionHandler jedisClusterConnectionHandler = jedisCluster.getConnectionHandler();
``

 -----------------------------OR -----------------------------
Change the visibility of getConnectionFromSlot(int slot) into JedisCluster and expose method into BinaryJedisCluster class same as getClusterNodes() method

`
 public Jedis getConnectionFromSlot(int slot) {
	return  this.connectionHandler.getConnectionFromSlot(slot);
 }
`
By this way no one can mess up connection handler and  we can get the Jedis Connection by slot without exposing connection handler.

### Redis / Jedis Configuration
We are using redis 3.2 cluster nodes
#### Jedis version:
2.8.2
#### Redis version:
3.2
#### Java version:
1.8

I request you kindly provide some solution overcome overloading of class issue. It is more appreciable if we can able to get reference of connection handler then every one can able to use pipeline on redis cluster.

Kindly find the google-group link, where i already raised the thread for same.
https://groups.google.com/forum/#!topic/jedis_redis/rK2G6F4FyuE

**Comment by saurabhygk** [comment]
Please check the scenario, Why I have implemented pipeline by my self. And Why I suggest to add the method into BinaryJedisCluster class.
[ConnectionHandler_Changes.docx](https://github.com/xetorthio/jedis/files/1086905/ConnectionHandler_Changes.docx)

**Comment by marcosnils** [comment]
@saurabhygk we don't like exposing the connectionHandler to avoid users mess with it. According to your use case seems like you can use the getConnectionFromSlot method (https://github.com/xetorthio/jedis/blob/9577fbdc8cecd6c8cd90c15a396683379b9883ee/src/main/java/redis/clients/jedis/JedisSlotBasedConnectionHandler.java#L64-L82) for that purse. Will exposing this function suit your needs?

**Comment by saurabhygk** [comment]
@marcosnils Thank you for replying me. I am using the getConnectionFromSlot method to get the connection for slot. But in my use case I am calling pipeline on Jedis Cluster but I can't do it because lack of feature. So we decided to get Jedis Connection using slot but for that we require connectionHandler and this function which i suggested exposing connectionHandler fulfils our needs. By connectionHandler we can able to get Jedis connection from jedis pool by using getConnectionFromSlot method and we can able to do pipeline on it.

**Comment by marcosnils** [comment]
> I am using the getConnectionFromSlot method to get the connection for slot.

You can already use pipeline on the connection that you get from `getConnectionFromSlot`. I still don't understand why you need the connectionHandler to be exposed.

**Comment by saurabhygk** [comment]
@marcosnils sorry I am not able to explain you. But the thing is we are using JedisClusterConnection and who does not provides the pipeline feature. 

To do pipeline operation on dedicated from Dedicated node we require Jedis connection right. Now to get Jedis Connection we can call getConnectionFromSlot this method by slot no, I agree. 

`JedisSlotBasedConnectionHandler` class provides `getConnectionFromSlot` this method. But to call this method we require `JedisSlotBasedConnectionHandler` object reference, and  JedisClusterConnection or JedisCluster not providing reference object.

Also check my code
`
		JedisClusterConnection jedisClusterConnection = (JedisClusterConnection) stringRedisTemplate.getConnectionFactory().getClusterConnection();
		JedisCluster jedisCluster = jedisClusterConnection.getNativeConnection();
		JedisSlotBasedConnectionHandler jedisClusterConnectionHandler = jedisCluster.getConnectionHandler(); // this is my exposed method
	 // can you please help me, without connectionHandler how I can get Jedis connection?
		Map<Integer, String> slotKeysMap = new HashMap<>();
		for (String key : keys) {
			]int slot = JedisClusterCRC16.getSlot(key);
			Jedis conn = jedisClusterConnectionHandler.getConnectionFromSlot(slot);
      Pipeline pipeline = conn.pipelined();
		}
`

If you want to check my use case please check attached document.
[ConnectionHandler_Changes.docx](https://github.com/xetorthio/jedis/files/1090079/ConnectionHandler_Changes.docx)

**Comment by marcosnils** [comment]
@saurabhygk I get what you're saying. What I'm proposing is to change `getConnectionFromSlot` visibility and expose it through JedisCluster, just like `getClusterNodes` (https://github.com/xetorthio/jedis/blob/master/src/main/java/redis/clients/jedis/BinaryJedisCluster.java#L74). Will this work for you?

**Comment by saurabhygk** [comment]
@marcosnils May be I have proposed same thing If i am not wrong. I tried to expose through JedisCluster by changing visibility in BinaryJedisCluster as you mentioned for getClusterNodes. 
Kindly check PR for same expose: https://github.com/xetorthio/jedis/pull/1528

**Comment by saurabhygk** [comment]
@marcosnils I can understand library do not want to expose connection handler. 

But yes we can change the visibility of getConnectionFromSlot(slotNo) method into JedisCluster by exposing it in BinaryJedisCluster class same as getClusterNodes method. 

I did Unit testing on my local machine by downloading Jedis 2.8.2 source code into local machine and exposed following method into BinaryJedisCluster class  for my scenario. This will be useful to other developer or community as well.

`
public Jedis getConnectionFromSlot(int slot) { 
return this.connectionHandler.getConnectionFromSlot(slot); 
}
`
So now I am planning to create New Pull Request for this change and will close existing one because that change is related to expose of ConnectionHandler reference object itself.

Existing Pull Request  : https://github.com/xetorthio/jedis/pull/1528 --> Closed by me because it is not relevant
New Pull Request :  https://github.com/xetorthio/jedis/pull/1532

**Comment by HeartSaVioR** [comment]
@saurabhygk 
I intentionally didn't implement JedisCluster with Pipeline because it is really hard to guarantee safeness. You should think about the scenario: `target slot moves while commands are sent in pipeline`.
Please refer below discussion from Redis group:
https://groups.google.com/d/msg/redis-db/4I0ELYnf3bk/Lrctk0ULm6AJ

Just like to say you might have hard time guaranteeing safeness, or might meet corner cases unless you can understand the discussion and provide a solution around the limitation.

**Comment by saurabhygk** [comment]
@HeartSaVioR I checked the group discussion and read your approach. I can understand the situation. Can you please share me your approach if possible, if it's implemented by you. else i will try to find out solution to overcome with this issue. Thank you for this concern

**Comment by HeartSaVioR** [comment]
The Josiah's suggestion in discussion would be safe:

```
To repeat again: wrap your batch of commands in MULTI/EXEC. While that won't prevent slot migration, no migration happens during the MULTI/EXEC execution, and all keys are checked before execution of any commands inside the MULTI/EXEC.
```

In fact, that is not the true-pipelining. Commands are sent to the Redis in pipelining manner, but commands will be executed only when Redis receives EXEC, hence like a batch. If slot moving occurs in the middle of pipelining, whole commands would be fail and you need to handle the situation well. (maybe want to retry to the node which is slot's new owner automatically)

**Comment by lzh0379** [comment]
Hi @HeartSaVioR, my theory of cluster pipelining is simple.

Suppose we have a batch of commands (request) sending to the cluster.
We can find the right connection by the key's check sum (CRC16), and then group those requests by connections. On each connection, we can send requestsi n pipeline mode.
Later, we just need read results from those connections in pipeline mode.

But node migration need to be considered, some operations will fail due to node migration.
We can know the right connection where the requests should go by checking the returned value.
We can replay the failed requests, just like we do it in the first round.

We can repeat this process, until all requests succeed.

For replaying, we need save the key and data of each request, we also need update the slot mapping wiith the server response.

Because we need wait for responces from multiple connections, this implementation would be less efficient than it on single connection.

**Comment by sdncha** [comment]
Hi @HeartSaVioR, my theory of cluster pipelining is simple.

Suppose we have a batch of commands (request) sending to the cluster.
We can find the right connection by the key's check sum (CRC16), and then group those requests by connections. On each connection, we can send requestsi n pipeline mode.
Later, we just need read results from those connections in pipeline mode.

But node migration need to be considered, some operations will fail due to node migration.
We can know the right connection where the requests should go by checking the returned value.
We can replay the failed requests, just like we do it in the first round.

We can repeat this process, until all requests succeed.

For replaying, we need save the key and data of each request, we also need update the slot mapping wiith the server response.

Because we need wait for responces from multiple connections, this implementation would be less efficient than it on single connection.

**Comment by sdncha** [comment]
Hi @HeartSaVioR, the replay is not really replay, it is just retrying the failed operations. So it is not limited to idempotent operations.

**Comment by sazzad16** [comment]
Resolved by #1532
